File: hellojava/src/main/java/com/fathzer/chess/optimized/IntList.java
package com.fathzer.chess.optimized;

/**
 * A resizable array implementation for primitive int values.
 * This class provides similar functionality to ArrayList<Integer> but avoids autoboxing overhead.
 */
public class IntList {
    private static final int DEFAULT_CAPACITY = 64;
    private int[] elements;
    private int size;

    /**
     * Constructs an empty list with an initial capacity of ten.
     */
    public IntList() {
        this.elements = new int[DEFAULT_CAPACITY];
        this.size = 0;
    }
    
    /**
     * Constructs a new IntList containing the elements of the specified IntList.
     * The new list will be a shallow copy of the original list.
     *
     * @param other the IntList whose elements are to be placed into this list
     * @throws NullPointerException if the specified IntList is null
     */
    public IntList(IntList other) {
        this.size = other.size;
        this.elements = new int[Math.max(size, DEFAULT_CAPACITY)];
        System.arraycopy(other.elements, 0, this.elements, 0, size);
    }

    /**
     * Appends the specified element to the end of this list.
     *
     * @param element element to be appended to this list
     */
    public void add(int element) {
        ensureCapacity(size + 1);
        elements[size++] = element;
    }
    
    /**
     * Appends all of the elements in the specified collection to the end of
     * this list, in the order that they are returned by the specified
     * collection's iterator.
     *
     * @param other collection containing elements to be added to this list
     * @throws NullPointerException if the specified collection is null
     */
    public void addAll(IntList other) {
        int numNew = other.size;
        ensureCapacity(size + numNew);
        System.arraycopy(other.elements, 0, elements, size, numNew);
        size += numNew;
    }

    /**
     * Removes and returns the last element from this list.
     *
     * @return the element that was removed from the list
     * @throws IllegalStateException if the list is empty
     */
    public int removeLast() {
        return elements[--size];
    }

    /**
     * Returns the element at the specified position in this list.
     *
     * @param index index of the element to return
     * @return the element at the specified position in this list
     * @throws IndexOutOfBoundsException if the index is out of range (index < 0 || index >= size())
     */
    public int get(int index) {
        return elements[index];
    }

    /**
     * Returns the number of elements in this list.
     *
     * @return the number of elements in this list
     */
    public int size() {
        return size;
    }

    /**
     * Removes all of the elements from this list. The list will be empty after this call returns.
     */
    public void clear() {
        size = 0;
    }
    
    /**
     * Returns true if this list contains the specified element.
     *
     * @param value element whose presence in this list is to be tested
     * @return true if this list contains the specified element, false otherwise
     */
    public boolean contains(int value) {
        for (int i = 0; i < size; i++) {
            if (elements[i] == value) {
                return true;
            }
        }
        return false;
    }

    /**
     * Increases the capacity of this IntList instance, if necessary, to ensure
     * that it can hold at least the number of elements specified by the minimum
     * capacity argument.
     *
     * @param minCapacity the desired minimum capacity
     */
    private void ensureCapacity(int minCapacity) {
        if (minCapacity > elements.length) {
            int newCapacity = elements.length + (elements.length >> 1); // Grow by 50%
            if (newCapacity < minCapacity) {
                newCapacity = minCapacity;
            }
            int[] newElements = new int[newCapacity];
            System.arraycopy(elements, 0, newElements, 0, size);
            elements = newElements;
        }
    }
}

File: hellogo/pkg/chess/optimized/intlist.go
// Package optimized provides optimized data structures for chess engine implementation.
package optimized

// DefaultCapacity is the default initial capacity of an IntList.
const DefaultCapacity = 64

// IntList is a resizable array implementation for primitive int values.
// It provides similar functionality to a slice but with a more controlled API.
type IntList struct {
	elements []int
	size     int
}

// NewIntList creates a new empty IntList with the default capacity.
func NewIntList() *IntList {
	return &IntList{
		elements: make([]int, DefaultCapacity),
		size:     0,
	}
}

// NewIntListFrom creates a new IntList that is a copy of another IntList.
func NewIntListFrom(other *IntList) *IntList {
	if other == nil {
		panic("IntList: source cannot be nil")
	}

	// Create a new list with the same size as the other list
	newList := &IntList{
		elements: make([]int, max(other.size, DefaultCapacity)),
		size:     other.size,
	}
	// Copy the elements
	copy(newList.elements, other.elements[:other.size])
	return newList
}

// Add appends the specified element to the end of this list.
func (l *IntList) Add(element int) {
	l.ensureCapacity(l.size + 1)
	l.elements[l.size] = element
	l.size++
}

// AddAll appends all elements from another IntList to this list.
func (l *IntList) AddAll(other *IntList) {
	if other == nil {
		return
	}

	numNew := other.size
	l.ensureCapacity(l.size + numNew)
	copy(l.elements[l.size:], other.elements[:numNew])
	l.size += numNew
}

// RemoveLast removes and returns the last element from this list.
func (l *IntList) RemoveLast() int {
	if l.size == 0 {
		panic("IntList: cannot remove from empty list")
	}

	l.size--
	return l.elements[l.size]
}

// Get returns the element at the specified position in this list.
func (l *IntList) Get(index int) int {
	if index < 0 || index >= l.size {
		panic("IntList: index out of bounds")
	}
	return l.elements[index]
}

// Size returns the number of elements in this list.
func (l *IntList) Size() int {
	return l.size
}

// Clear removes all elements from this list.
func (l *IntList) Clear() {
	l.size = 0
}

// Contains returns true if this list contains the specified value.
func (l *IntList) Contains(value int) bool {
	for i := 0; i < l.size; i++ {
		if l.elements[i] == value {
			return true
		}
	}
	return false
}

// ensureCapacity ensures that the list can hold at least minCapacity elements.
func (l *IntList) ensureCapacity(minCapacity int) {
	if minCapacity > len(l.elements) {
		newCapacity := len(l.elements) + (len(l.elements) >> 1) // Grow by 50%
		if newCapacity < minCapacity {
			newCapacity = minCapacity
		}

		newElements := make([]int, newCapacity)
		copy(newElements, l.elements[:l.size])
		l.elements = newElements
	}
}

// ToSlice returns a new slice containing all elements in this list.
func (l *IntList) ToSlice() []int {
	slice := make([]int, l.size)
	copy(slice, l.elements[:l.size])
	return slice
}
